README
USB CCID IFD Handler
====================

This package provides the source code for a generic USB CCID (Chip/Smart
Card Interface Devices) and ICCD (Integrated Circuit(s) Card Devices)
driver. See the USB CCID [1] and ICCD [2] specifications from the USB
working group.

[1] http://www.usb.org/developers/devclass_docs/DWG_Smart-Card_CCID_Rev110.pdf
[2] http://www.usb.org/developers/devclass_docs/DWG_Smart-Card_USB-ICC_ICCD_rev10.pdf

Authors:
========

- Ludovic Rousseau <ludovic.rousseau@free.fr>
- Carlos Prados for the PPS and ATR parsing code (taken from his
  towitoto driver) in towitoko/ directory.
- Olaf Kirch for the T=1 TPDU code (from the OpenCT package) in openct/
  directory. I (Ludovic Rousseau) greatly improved this code.


CCID and ICCD readers:
======================

A reader can be in one of these list:
- supported
    See http://pcsclite.alioth.debian.org/ccid/supported.html
- should working
    See http://pcsclite.alioth.debian.org/ccid/shouldwork.html
- unsupported
    See http://pcsclite.alioth.debian.org/ccid/unsupported.html
- disabled
    See http://pcsclite.alioth.debian.org/ccid/disabled.html


Supported operating systems:
============================

- GNU/Linux (libusb 1.0)
- MacOS X/Darwin (libusb 1.0)

See also http://pcsclite.alioth.debian.org/ccid.html for more
information.


Debug informations:
===================

The driver uses the debug function provided by pcscd. So if pcscd sends
its debug to stdout (pcscd --foreground) then the CCID driver will also
send its debug to stdout. If pcscd sends its debug to syslog (by
default) then the CCID driver will also send its debug to syslog.

You can change the debug level using the Info.plist configuration file.
The Info.plist is installed, by default, in
/usr/local/pcsc/drivers/ifd-ccid.bundle/Contents/Info.plist
or set the environment variable LIBCCID_ifdLogLevel.

The debug level is set in the ifdLogLevel field. It is a binary OR
combinaison of 4 different levels.
- 1: critical: important error messages
- 2: info:     informative messages like what reader was detected
- 4: comm:     a dump of all the bytes exchanged between the host and the
               reader
- 8: periodic: periodic info when pcscd test if a card is present (every
               1/10 of a second)

By default the debug level is set to 3 (1 + 2) and correspond to the
critical and info levels.

You have to restart the driver so it read the configuration file again
and use the new debug level value.  To restart the driver you just need
to unplug all your CCID readers so the the driver is unloaded and then
replug your readers. You can also restart pcscd.


Voltage selection
=================

You can change the voltage level using the Info.plist configuration
file.  The Info.plist is installed, by default, in
/usr/local/pcsc/drivers/ifd-ccid.bundle/Contents/Info.plist

The voltage level is set in the ifdDriverOptions field. It is a binary OR
combinaison of 4 different levels.

-  0: power on the card at 5V (default value)
- 16: power on the card at 3V and, if 3V fails then use 5V
- 32: power on the card at 1.8V, then 3V and then 5V
- 48: let the reader decide

By default the voltage level is set to 0 and correspond to 5V.

You have to restart the driver so it read the configuration file again
and use the new debug level value.  To restart the driver you just need
to unplug all your CCID readers so the the driver is unloaded and then
replug your readers.  You can also restart pcscd.


Licence:
========

  This library is free software; you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or (at
your option) any later version.

  This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser
General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
along with this library; if not, write to the Free Software Foundation,
Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


History:
========

1.4.15 - 14 February 2014, Ludovic Rousseau
    - Add support of
      . DUALi DRAGON NFC READER
      . Feitian bR301
      . Gemalto CR30 reader in serial communication
      . Gemalto Ezio Shield Pro SC
      . IIT E.Key Almaz-1C
    - PIN_MODIFY_STRUCTURE & PIN_VERIFY_STRUCTURE: Fix calculation of
      the command length after pcsc-lite 1.8.9 (October 2013) changed
      the PCSC/reader.h header
    - Add specific PIN min (0) & max (25) sizes for SmartTerminal
      ST-2xxx
    - Do not get the data rates if bNumDataRatesSupported = 0
    - Support Gemalto features for pinpad readers MinimumPINSize,
      MaximumPINSize and bEntryValidationCondition are fetched from the
      reader firmware
    - disable (broken) pinpad for Fujitsu SmartCase KB SCR eSIG
    - examples/scardcontrol.c:
      . Parse codes returned by a pinpad (as SW1/SW2)
        Known codes for now are:
        0x9000: Success
        0x6400: Timeout
        0x6401: Cancelled by user
        0x6402: PIN mismatch
        0x6403: Too short or too long PIN
      . Retrieve min and max PIN sizes from the driver
      . Retrieve bEntryValidationCondition from the driver
    - be more strict for bInterfaceClass = 255 by also checking extra_length
    - some minor bugs removed


1.4.14 - 25 November 2013, Ludovic Rousseau
    - Add support of
      . Gemalto GemCore SIM Pro firmware 2.0 (using USB)
    - report FEATURE_IFD_PIN_PROPERTIES only for pinpad readers
    - Generalize the management of (old) readers with bDeviceClass = 0xFF
    - some minor bugs removed


1.4.13 - 9 October 2013, Ludovic Rousseau
    - Add support of
      . Access IS ePassport Reader
      . Planeta RC700-NFC CCID
    - Add support of Windows value for CM_IOCTL_GET_FEATURE_REQUEST
      Windows uses 0x313520 for SCARD_CTL_CODE(3400) pcsc-lite uses
      0x42000D48 for SCARD_CTL_CODE(3400)
      RDP aplications (like rdesktop) will convert SCardControl()
      commands from a Windows application (so using 0x313520) to
      pcsc-lite.
    - fix multi-slot support for card movement notification (introduced
      in 1.4.12)
    - Mac OS X: differentiate each libccid library by the dynamic linker
      using --prefix=/fake/$BUNDLE_ID
    - some minor bugs removed


1.4.12 - 12 August 2013, Ludovic Rousseau
    - Add support of
      . HID OMNIKEY 5127 CK
      . HID OMNIKEY 5326 DFR
      . HID OMNIKEY 5427 CK
      . Ingenico WITEO USB Smart Card Reader (Base and Badge)
      . SecuTech SecuTech Token
    - Add support of card movement notifications for multi-slot readers
    - Check libusb is at least at version 1.0.8
    - Get the serialconfdir value from pcsc-lite pkg config instead of
      using $(DESTDIR)/$(sysconfdir)/reader.conf.d/
    - Disable class driver on Mac OS X
    - Update the bundle name template to include the vendor name
    - some minor bugs removed


1.4.11 - 12 June 2013, Ludovic Rousseau
    - Add support of
       . Gemalto IDBridge CT30
       . Gemalto IDBridge K30
       . SCM Microsystems Inc. SCL010 Contactless Reader
       . SCM Microsystems Inc. SDI011 Contactless Reader
       . THRC reader
    - Better management of time extension requests
    - parse: better support of devices with bInterfaceClass = 0xFF
    - udev rule file: Remove setting group to pcscd, remove support of
      Linux kernel < 2.6.35 for auto power up management
    - some minor bugs removed


1.4.10 - 16 April 2013, Ludovic Rousseau
    - Add support of
       . ACS APG8201 USB Reader with PID 0x8202
       . GIS Ltd SmartMouse USB
       . Gemalto IDBridge K3000
       . Identive CLOUD 2700 F Smart Card Reader
       . Identive CLOUD 2700 R Smart Card Reader
       . Identive CLOUD 4500 F Dual Interface Reader
       . Identive CLOUD 4510 F Contactless + SAM Reader
       . Identive CLOUD 4700 F Dual Interface Reader
       . Identive CLOUD 4710 F Contactless + SAM Reader
       . Inside Secure AT90SCR050
       . Inside Secure AT90SCR100
       . Inside Secure AT90SCR200
       . SCR3310-NTTCom USB SmartCard Reader
       . SafeTech SafeTouch
       . SpringCard H512 Series
       . SpringCard H663 Series
       . SpringCard NFC'Roll
       . Yubico Yubikey NEO CCID
       . Yubico Yubikey NEO OTP+CCID
    - Add support of time extension for Escape commands


1.4.9 - 16 January 2013, Ludovic Rousseau
    - Add support of
       . Aktiv Rutoken PINPad In
       . Aktiv Rutoken PINPad Ex
       . REINER SCT cyberJack go
    - Info.plist: Correctly handle reader names containing &


1.4.8 - 22 June 2012, Ludovic Rousseau
    - Add support of
       . SCR3310-NTTCom USB (was removed in version 1.4.6)
       . Inside Secure VaultIC 420 Smart Object
       . Inside Secure VaultIC 440 Smart Object
    - Wait up to 3 seconds for reader start up
    - Add support of new PC/SC V2 part 10 properties:
        . dwMaxAPDUDataSize
        . wIdVendor
        . wIdProduct
    - Use helper functions from libPCSCv2part10 to parse the PC/SC v2
      part 10 features


1.4.7 - 22 June 2012, Ludovic Rousseau
    - Add support of
       . ACS ACR101 ICC Reader
       . ACS CryptoMate64
       . Alcor Micro AU9522
       . Bit4id CKey4
       . Bit4id cryptokey
       . Bit4id iAM
       . Bit4id miniLector
       . Bit4id miniLector-s
       . CCB eSafeLD
       . Gemalto Ezio Shield Branch
       . KOBIL Systems IDToken
       . NXP PR533
    - KOBIL Systems IDToken special cases:
       . Give more time (3 seconds instead of 2) to the reader to answer
       . Hack for the Kobil IDToken and Geman eID card. The German eID
         card is bogus and need to be powered off before a power on
       . Add Reader-Info-Commands special APDU/command
         - Manufacturer command
         - Product name command
         - Firmware version command
         - Driver version command
    - Use auto suspend for CCID devices only (Closes Alioth bug
      [#313445] "Do not activate USB suspend for composite devices:
      keyboard")
    - Fix some error management in the T=1 TPDU state machine
    - some minor bugs removed
    - some minor improvements added


1.4.6 - 6 April 2012, Ludovic Rousseau
    - Add support of
       . Avtor SC Reader 371
       . Avtor SecureToken
       . DIGIPASS KEY 202
       . Fujitsu SmartCase KB SCR eSIG
       . Giesecke & Devrient StarSign CUT
       . Inside Secure VaultIC 460 Smart Object
       . Macally NFC CCID eNetPad reader
       . OmniKey 6321 USB
       . SCM SDI 011
       . Teridian TSC12xxF
       . Vasco DIGIPASS KEY 101
    - Remove support of readers without a USB CCID descriptor file
       . 0x08E6:0x34C1:Gemalto Ezio Shield Secure Channel
       . 0x08E6:0x34C4:Gemalto Ezio Generic
       . 0x04E6:0x511A:SCM SCR 3310 NTTCom
       . 0x0783:0x0008:C3PO LTC32 USBv2 with keyboard support
       . 0x0783:0x9002:C3PO TLTC2USB
       . 0x047B:0x020B:Silitek SK-3105
    - Disable SPE for HP USB CCID Smartcard Keyboard. The reader is
      bogus and unsafe.
    - Convert "&" in a reader name into "&amp;" to fix a problem on Mac
      OS X
    - Fix a problem with ICCD type A devices. We now wait for device ready
    - Secure PIN Verify and PIN Modify: set the minimum timeout to 90
      seconds
    - Add support of wIdVendor and wIdProduct properties
    - Add support of dwMaxAPDUDataSize
    - Add support of Gemalto firmware features
    - some minor bugs removed


1.4.5 - 11 October 2011, Ludovic Rousseau
    - Add support of
      . Alcor Micro AU9540
      . BIFIT USB-Token iBank2key
      . BIFIT iBank2Key
      . Gemalto Ezio Shield PinPad reader
      . Gemalto SA .NET Dual
      . Precise Sense MC reader (with fingerprint)
      . SDS DOMINO-Key TWIN Pro
      . Ubisys 13.56MHz RFID (CCID)
    - Add support of bPPDUSupport and FEATURE_CCID_ESC_COMMAND
    - SCARD_ATTR_VENDOR_NAME and SCARD_ATTR_VENDOR_IFD_VERSION are not
      the vendor name and version of the driver but of the IFD:
      InterFace Device i.e. the smart card reader.  We then return the
      USB iManufacturer string as SCARD_ATTR_VENDOR_NAME and USB
      bcdDevice as SCARD_ATTR_VENDOR_IFD_VERSION
    - reduce binary size bu removing unused features from simclist
    - Fix some warnings reported bu Coverity


1.4.4 - 13 May 2011, Ludovic Rousseau
    - Add support of
      . Gemalto Ezio CB+
      . Gemalto Ezio Generic
      . Gemalto Ezio Shield
      . Gemalto Ezio Shield PinPad
      . Gemalto Ezio Shield Secure Channel
    - Activate USB automatic power suspend. The Linux kernel should
      power off the reader automatically if it is not used (pcscd is not
      running).
    - Add support of TLV Properties wLcdMaxCharacters and wLcdMaxLines.
      They just duplicate wLcdLayout
    - some minor bugs removed


1.4.3 - 2 April 2011, Ludovic Rousseau
    - Add support of
      . Feitian ePass2003 readers
      . Neowave Weneo
      . SCM SCL011
      . Vasco DIGIPASS 920
    - use :libudev: instead of :libhal: naming scheme.
    - Do not install RSA_SecurID_getpasswd and Kobil_mIDentity_switch
      and the associated documentation.
    - the Secure Pin Entry of the HP USB Smart Card Keyboard is bogus so
      disable it
    - some minor bugs removed


1.4.2 - 22 February 2011, Ludovic Rousseau
    - Add support of
      . ACS APG8201 PINhandy 1
      . Aktiv Rutoken lite readers
      . Feitian SCR310 reader (also known as 301v2)
      . Oberthur ID-ONE TOKEN SLIM v2
      . Vasco DIGIPASS KEY 200
      . Vasco DIGIPASS KEY 860
      . Xiring Leo v2
      . Xiring MyLeo
      . new Neowave Weneo token
    - Add back support of "bogus" Oz776, REINER SCT and BLUDRIVE II
    - Ease detection of OpenCT by pcsc-lite
    - disable use of interrupt card events for multi slots readers (the
      algorithm is bogus and can't be used)
    - fix minor problems detected by the clang tool
    - some minor bugs removed


1.4.1 - 3 December 2010, Ludovic Rousseau
    - Add support of
      . Akasa AK-CR-03, BZH uKeyCI800-K18
      . Free Software Initiative of Japan Gnuk token readers
      . Gemalto Smart Guardian (SG CCID)
      . ReinerSCT cyberJack RFID basis
    - Remove O2 Micro Oz776 and Blutronics Bludrive II CCID since they
      are no more supported since version 1.4.0
    - SecurePINVerify() & SecurePINModify(): Accept big and little
      endian byte orders for multibytes fields. The application
      should not use HOST_TO_CCID_16() and HOST_TO_CCID_32() any more
      and just use the normal byte order of the architecture.
    - Need pcsc-lite 1.6.5 for TAG_IFD_POLLING_THREAD_WITH_TIMEOUT

NEWS
Read the README file for news.


src/ccid.c
/*
    ccid.c: CCID common code
    Copyright (C) 2003-2010   Ludovic Rousseau

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

	You should have received a copy of the GNU Lesser General Public License
	along with this library; if not, write to the Free Software Foundation,
	Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

/*
 * $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <pcsclite.h>
#include <ifdhandler.h>

#include "config.h"
#include "debug.h"
#include "ccid.h"
#include "defs.h"
#include "ccid_ifdhandler.h"
#include "commands.h"
#include "utils.h"

#ifdef __APPLE__
#include <CoreFoundation/CoreFoundation.h>
#endif

/*****************************************************************************
 *
 *					ccid_open_hack_pre
 *
 ****************************************************************************/
int ccid_open_hack_pre(unsigned int reader_index)
{
	_ccid_descriptor *ccid_descriptor = get_ccid_descriptor(reader_index);

	switch (ccid_descriptor->readerID)
	{
		case CARDMAN3121+1:
			/* Reader announces APDU but is in fact TPDU */
			ccid_descriptor->dwFeatures &= ~CCID_CLASS_EXCHANGE_MASK;
			ccid_descriptor->dwFeatures |= CCID_CLASS_TPDU;
			break;

		case MYSMARTPAD:
			ccid_descriptor->dwMaxIFSD = 254;
			break;

		case CL1356D:
			/* the firmware needs some time to initialize */
			(void)sleep(1);
			ccid_descriptor->readTimeout = 60*1000; /* 60 seconds */
			break;

		case GEMPCTWIN:
		case GEMPCKEY:
		case DELLSCRK:
			/* Only the chipset with firmware version 2.00 is "bogus"
			 * The reader may send packets of 0 bytes when the reader is
			 * connected to a USB 3 port */
			if (0x0200 == ccid_descriptor->IFD_bcdDevice)
			{
				ccid_descriptor->zlp = TRUE;
				DEBUG_INFO("ZLP fixup");
			}
			break;
	}

	/* CCID */
	if ((PROTOCOL_CCID == ccid_descriptor->bInterfaceProtocol)
		&& (3 == ccid_descriptor -> bNumEndpoints))
	{
#ifndef TWIN_SERIAL
		/* just wait for 100ms in case a notification is in the pipe */
		(void)InterruptRead(reader_index, 100);
#endif
	}

	/* ICCD type A */
	if (PROTOCOL_ICCD_A == ccid_descriptor->bInterfaceProtocol)
	{
		unsigned char tmp[MAX_ATR_SIZE];
		unsigned int n = sizeof(tmp);

		DEBUG_COMM("ICCD type A");
		(void)CmdPowerOff(reader_index);
		(void)CmdPowerOn(reader_index, &n, tmp, CCID_CLASS_AUTO_VOLTAGE);
		(void)CmdPowerOff(reader_index);
	}

	/* ICCD type B */
	if (PROTOCOL_ICCD_B == ccid_descriptor->bInterfaceProtocol)
	{
		unsigned char tmp[MAX_ATR_SIZE];
		unsigned int n = sizeof(tmp);

		DEBUG_COMM("ICCD type B");
		if (CCID_CLASS_SHORT_APDU ==
			(ccid_descriptor->dwFeatures & CCID_CLASS_EXCHANGE_MASK))
		{
			/* use the extended APDU comm alogorithm */
			ccid_descriptor->dwFeatures &= ~CCID_CLASS_EXCHANGE_MASK;
			ccid_descriptor->dwFeatures |= CCID_CLASS_EXTENDED_APDU;
		}

		(void)CmdPowerOff(reader_index);
		(void)CmdPowerOn(reader_index, &n, tmp, CCID_CLASS_AUTO_VOLTAGE);
		(void)CmdPowerOff(reader_index);
	}

	return 0;
} /* ccid_open_hack_pre */

#ifndef NO_LOG
/*****************************************************************************
 *
 *					dump_gemalto_firmware_features
 *
 ****************************************************************************/
static void dump_gemalto_firmware_features(struct GEMALTO_FIRMWARE_FEATURES *gff)
{
	DEBUG_INFO2("Dumping Gemalto firmware features (%zd bytes):",
		sizeof(struct GEMALTO_FIRMWARE_FEATURES));

#define YESNO(x) (x) ? "yes" : "no"

	DEBUG_INFO2(" bLogicalLCDLineNumber: %d", gff->bLogicalLCDLineNumber);
	DEBUG_INFO2(" bLogicalLCDRowNumber: %d", gff->bLogicalLCDRowNumber);
	DEBUG_INFO2(" bLcdInfo: 0x%02X", gff->bLcdInfo);
	DEBUG_INFO2(" bEntryValidationCondition: 0x%02X",
		gff->bEntryValidationCondition);

	DEBUG_INFO(" Reader supports PC/SCv2 features:");
	DEBUG_INFO2("  VerifyPinStart: %s", YESNO(gff->VerifyPinStart));
	DEBUG_INFO2("  VerifyPinFinish: %s", YESNO(gff->VerifyPinFinish));
	DEBUG_INFO2("  ModifyPinStart: %s", YESNO(gff->ModifyPinStart));
	DEBUG_INFO2("  ModifyPinFinish: %s", YESNO(gff->ModifyPinFinish));
	DEBUG_INFO2("  GetKeyPressed: %s", YESNO(gff->GetKeyPressed));
	DEBUG_INFO2("  VerifyPinDirect: %s", YESNO(gff->VerifyPinDirect));
	DEBUG_INFO2("  ModifyPinDirect: %s", YESNO(gff->ModifyPinDirect));
	DEBUG_INFO2("  Abort: %s", YESNO(gff->Abort));
	DEBUG_INFO2("  GetKey: %s", YESNO(gff->GetKey));
	DEBUG_INFO2("  WriteDisplay: %s", YESNO(gff->WriteDisplay));
	DEBUG_INFO2("  SetSpeMessage: %s", YESNO(gff->SetSpeMessage));
	DEBUG_INFO2("  bTimeOut2: %s", YESNO(gff->bTimeOut2));
	DEBUG_INFO2("  bPPDUSupportOverXferBlock: %s",
		YESNO(gff->bPPDUSupportOverXferBlock));
	DEBUG_INFO2("  bPPDUSupportOverEscape: %s",
		YESNO(gff->bPPDUSupportOverEscape));

	DEBUG_INFO2(" bListSupportedLanguages: %s",
		YESNO(gff->bListSupportedLanguages));
	DEBUG_INFO2(" bNumberMessageFix: %s", YESNO(gff->bNumberMessageFix));

	DEBUG_INFO2(" VersionNumber: 0x%02X", gff->VersionNumber);
	DEBUG_INFO2(" MinimumPINSize: %d", gff->MinimumPINSize);
	DEBUG_INFO2(" MaximumPINSize: %d", gff->MaximumPINSize);
	DEBUG_INFO2(" Firewall: %s", YESNO(gff->Firewall));
	if (gff->Firewall && gff->FirewalledCommand_SW1
		&& gff->FirewalledCommand_SW2)
	{
		DEBUG_INFO2("  FirewalledCommand_SW1: 0x%02X",
			gff->FirewalledCommand_SW1);
		DEBUG_INFO2("  FirewalledCommand_SW2: 0x%02X",
			gff->FirewalledCommand_SW2);
	}

} /* dump_gemalto_firmware_features */
#endif

/*****************************************************************************
 *
 *					set_gemalto_firmware_features
 *
 ****************************************************************************/
static void set_gemalto_firmware_features(unsigned int reader_index)
{
	_ccid_descriptor *ccid_descriptor = get_ccid_descriptor(reader_index);
	struct GEMALTO_FIRMWARE_FEATURES *gf_features;

	gf_features = malloc(sizeof(struct GEMALTO_FIRMWARE_FEATURES));
	if (gf_features)
	{
		unsigned char cmd[] = { 0x6A }; /* GET_FIRMWARE_FEATURES command id */
		unsigned int len_features = sizeof *gf_features;
		RESPONSECODE ret;

		ret = CmdEscape(reader_index, cmd, sizeof cmd,
			(unsigned char*)gf_features, &len_features, 0);
		if ((IFD_SUCCESS == ret) &&
		    (len_features == sizeof *gf_features))
		{
			/* Command is supported if it succeeds at CCID level */
			/* and returned size matches our expectation */
			ccid_descriptor->gemalto_firmware_features = gf_features;
#ifndef NO_LOG
			dump_gemalto_firmware_features(gf_features);
#endif
		}
		else
		{
			/* Command is not supported, let's free allocated memory */
			free(gf_features);
			DEBUG_INFO3("GET_FIRMWARE_FEATURES failed: " DWORD_D ", len=%d",
				ret, len_features);
		}
	}
} /* set_gemalto_firmware_features */

/*****************************************************************************
 *
 *					ccid_open_hack_post
 *
 ****************************************************************************/
int ccid_open_hack_post(unsigned int reader_index)
{
	_ccid_descriptor *ccid_descriptor = get_ccid_descriptor(reader_index);
	RESPONSECODE return_value = IFD_SUCCESS;

	switch (ccid_descriptor->readerID)
	{
		case GEMPCKEY:
		case GEMPCTWIN:
			/* Reader announces TPDU but can do APDU (EMV in fact) */
			if (DriverOptions & DRIVER_OPTION_GEMPC_TWIN_KEY_APDU)
			{
				unsigned char cmd[] = { 0x1F, 0x02 };
				unsigned char res[10];
				unsigned int length_res = sizeof(res);

				if (CmdEscape(reader_index, cmd, sizeof(cmd), res, &length_res, 0) == IFD_SUCCESS)
				{
					ccid_descriptor->dwFeatures &= ~CCID_CLASS_EXCHANGE_MASK;
					ccid_descriptor->dwFeatures |= CCID_CLASS_SHORT_APDU;
				}
			}
			break;

		case VEGAALPHA:
		case GEMPCPINPAD:
			/* load the l10n strings in the pinpad memory */
			{
#define L10N_HEADER_SIZE 5
#define L10N_STRING_MAX_SIZE 16
#define L10N_NB_STRING 10

				unsigned char cmd[L10N_HEADER_SIZE + L10N_NB_STRING * L10N_STRING_MAX_SIZE];
				unsigned char res[20];
				unsigned int length_res = sizeof(res);
				int offset, i, j;

				const char *fr[] = {
					"Entrer PIN",
					"Nouveau PIN",
					"Confirmer PIN",
					"PIN correct",
					"PIN Incorrect !",
					"Delai depasse",
					"* essai restant",
					"Inserer carte",
					"Erreur carte",
					"PIN bloque" };

				const char *de[] = {
					"PIN eingeben",
					"Neue PIN",
					"PIN bestatigen",
					"PIN korrect",
					"Falsche PIN !",
					"Zeit abgelaufen",
					"* Versuche ubrig",
					"Karte einstecken",
					"Fehler Karte",
					"PIN blockiert" };

				const char *es[] = {
					"Introducir PIN",
					"Nuevo PIN",
					"Confirmar PIN",
					"PIN OK",
					"PIN Incorrecto !",
					"Tiempo Agotado",
					"* ensayos quedan",
					"Introducir Tarj.",
					"Error en Tarjeta",
					"PIN bloqueado" };

				const char *it[] = {
					"Inserire PIN",
					"Nuovo PIN",
					"Confermare PIN",
					"PIN Corretto",
					"PIN Errato !",
					"Tempo scaduto",
					"* prove rimaste",
					"Inserire Carta",
					"Errore Carta",
					"PIN ostruito"};

				const char *pt[] = {
					"Insira PIN",
					"Novo PIN",
					"Conf. novo PIN",
					"PIN OK",
					"PIN falhou!",
					"Tempo expirou",
					"* tentiv. restam",
					"Introduza cartao",
					"Erro cartao",
					"PIN bloqueado" };

				const char *nl[] = {
					"Inbrengen code",
					"Nieuwe code",
					"Bevestig code",
					"Code aanvaard",
					"Foute code",
					"Time out",
					"* Nog Pogingen",
					"Kaart inbrengen",
					"Kaart fout",
					"Kaart blok" };

				const char *tr[] = {
					"PIN Giriniz",
					"Yeni PIN",
					"PIN Onayala",
					"PIN OK",
					"Yanlis PIN",
					"Zaman Asimi",
					"* deneme kaldi",
					"Karti Takiniz",
					"Kart Hatasi",
					"Kart Kilitli" };

				const char *en[] = {
					"Enter PIN",
					"New PIN",
					"Confirm PIN",
					"PIN OK",
					"Incorrect PIN!",
					"Time Out",
					"* retries left",
					"Insert Card",
					"Card Error",
					"PIN blocked" };

				const char *lang;
				const char **l10n;

#ifdef __APPLE__
				CFArrayRef cfa;
				CFStringRef slang;

				/* Get the complete ordered list */
				cfa = CFLocaleCopyPreferredLanguages();

				/* Use the first/preferred language
				 * As the driver is run as root we get the language
				 * selected during install */
				slang = CFArrayGetValueAtIndex(cfa, 0);

				/* CFString -> C string */
				lang = CFStringGetCStringPtr(slang, kCFStringEncodingMacRoman);
#else
				/* The other Unixes just use the LANG env variable */
				lang = getenv("LANG");
#endif
				DEBUG_COMM2("Using lang: %s", lang);
				if (NULL == lang)
					l10n = en;
				else
				{
					if (0 == strncmp(lang, "fr", 2))
						l10n = fr;
					else if (0 == strncmp(lang, "de", 2))
						l10n = de;
					else if (0 == strncmp(lang, "es", 2))
						l10n = es;
					else if (0 == strncmp(lang, "it", 2))
						l10n = it;
					else if (0 == strncmp(lang, "pt", 2))
						l10n = pt;
					else if (0 == strncmp(lang, "nl", 2))
						l10n = nl;
					else if (0 == strncmp(lang, "tr", 2))
						l10n = tr;
					else
						l10n = en;
				}

#ifdef __APPLE__
				/* Release the allocated array */
				CFRelease(cfa);
#endif
				offset = 0;
				cmd[offset++] = 0xB2;	/* load strings */
				cmd[offset++] = 0xA0;	/* address of the memory */
				cmd[offset++] = 0x00;	/* address of the first byte */
				cmd[offset++] = 0x4D;	/* magic value */
				cmd[offset++] = 0x4C;	/* magic value */

				/* for each string */
				for (i=0; i<L10N_NB_STRING; i++)
				{
					/* copy the string */
					for (j=0; l10n[i][j]; j++)
						cmd[offset++] = l10n[i][j];

					/* pad with " " */
					for (; j<L10N_STRING_MAX_SIZE; j++)
						cmd[offset++] = ' ';
				}

				(void)sleep(1);
				if (IFD_SUCCESS == CmdEscape(reader_index, cmd, sizeof(cmd), res, &length_res, DEFAULT_COM_READ_TIMEOUT))
				{
					DEBUG_COMM("l10n string loaded successfully");
				}
				else
				{
					DEBUG_COMM("Failed to load l10n strings");
					return_value = IFD_COMMUNICATION_ERROR;
				}
			}
			break;

		case HPSMARTCARDKEYBOARD:
		case HP_CCIDSMARTCARDKEYBOARD:
		case FUJITSUSMARTKEYB:
			/* the Secure Pin Entry is bogus so disable it
			 * http://martinpaljak.net/2011/03/19/insecure-hp-usb-smart-card-keyboard/
			 */
			ccid_descriptor->bPINSupport = 0;
			break;

#if 0
		/* SCM SCR331-DI contactless */
		case SCR331DI:
		/* SCM SCR331-DI-NTTCOM contactless */
		case SCR331DINTTCOM:
		/* SCM SDI010 contactless */
		case SDI010:
			/* the contactless reader is in the second slot */
			if (ccid_descriptor->bCurrentSlotIndex > 0)
			{
				unsigned char cmd1[] = { 0x00 };
				/*  command: 00 ??
				 * response: 06 10 03 03 00 00 00 01 FE FF FF FE 01 ?? */
				unsigned char cmd2[] = { 0x02 };
				/*  command: 02 ??
				 * response: 00 ?? */

				unsigned char res[20];
				unsigned int length_res = sizeof(res);

				if ((IFD_SUCCESS == CmdEscape(reader_index, cmd1, sizeof(cmd1), res, &length_res, 0))
					&& (IFD_SUCCESS == CmdEscape(reader_index, cmd2, sizeof(cmd2), res, &length_res, 0)))
				{
					DEBUG_COMM("SCM SCR331-DI contactless detected");
				}
				else
				{
					DEBUG_COMM("SCM SCR331-DI contactless init failed");
				}

				/* hack since the contactless reader do not share dwFeatures */
				ccid_descriptor->dwFeatures &= ~CCID_CLASS_EXCHANGE_MASK;
				ccid_descriptor->dwFeatures |= CCID_CLASS_SHORT_APDU;

				ccid_descriptor->dwFeatures |= CCID_CLASS_AUTO_IFSD;
			}
			break;
#endif
	}

	/* Gemalto readers may report additional information */
	if (GET_VENDOR(ccid_descriptor->readerID) == VENDOR_GEMALTO)
		set_gemalto_firmware_features(reader_index);

	return return_value;
} /* ccid_open_hack_post */

/*****************************************************************************
 *
 *					ccid_error
 *
 ****************************************************************************/
void ccid_error(int error, const char *file, int line, const char *function)
{
	const char *text;
	char var_text[30];

	switch (error)
	{
		case 0x00:
			text = "Command not supported or not allowed";
			break;

		case 0x01:
			text = "Wrong command length";
			break;

		case 0x05:
			text = "Invalid slot number";
			break;

		case 0xA2:
			text = "Card short-circuiting. Card powered off";
			break;

		case 0xA3:
			text = "ATR too long (> 33)";
			break;

		case 0xAB:
			text = "No data exchanged";
			break;

		case 0xB0:
			text = "Reader in EMV mode and T=1 message too long";
			break;

		case 0xBB:
			text = "Protocol error in EMV mode";
			break;

		case 0xBD:
			text = "Card error during T=1 exchange";
			break;

		case 0xBE:
			text = "Wrong APDU command length";
			break;

		case 0xE0:
			text = "Slot busy";
			break;

		case 0xEF:
			text = "PIN cancelled";
			break;

		case 0xF0:
			text = "PIN timeout";
			break;

		case 0xF2:
			text = "Busy with autosequence";
			break;

		case 0xF3:
			text = "Deactivated protocol";
			break;

		case 0xF4:
			text = "Procedure byte conflict";
			break;

		case 0xF5:
			text = "Class not supported";
			break;

		case 0xF6:
			text = "Protocol not supported";
			break;

		case 0xF7:
			text = "Invalid ATR checksum byte (TCK)";
			break;

		case 0xF8:
			text = "Invalid ATR first byte";
			break;

		case 0xFB:
			text = "Hardware error";
			break;

		case 0xFC:
			text = "Overrun error";
			break;

		case 0xFD:
			text = "Parity error during exchange";
			break;

		case 0xFE:
			text = "Card absent or mute";
			break;

		case 0xFF:
			text = "Activity aborted by Host";
			break;

		default:
			if ((error >= 1) && (error <= 127))
				(void)snprintf(var_text, sizeof(var_text), "error on byte %d",
					error);
			else
				(void)snprintf(var_text, sizeof(var_text),
					"Unknown CCID error: 0x%02X", error);

			text = var_text;
			break;
	}
	log_msg(PCSC_LOG_ERROR, "%s:%d:%s %s", file, line, function, text);

} /* ccid_error */

